### PR TITLE
KEYCLOAK-6659 - Add pairwise sub support to authorization services

### PR DESCRIPTION
**Pull request by hardselius**
Resolves: KEYCLOAK-6659

### CODE REVIEW

File: services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java [hunk 1]
```diff
@@ -1082,10 +1082,16 @@ public static AuthResult verifyIdentityToken(KeycloakSession session, RealmModel
                 }
             }
 
-            UserModel user = session.users().getUserById(token.getSubject(), realm);
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java]: I'm wondering how pairwise was working without this change. It seems that token verification would always fail. Is my understanding correct ?

Comment (hardselius) [reply services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java]: I'm not sure that part of the code is involved when verifying refresh tokens. But you're right. It's odd.

**Comment by pedroigor** [comment]
@hardselius, it seems we have some failures. Just triggered a new build expecting a better result.

If it continues to fail, I can give it a try and see what is happening. If you are not doing this already :)

**Comment by hardselius** [comment]
@pedroigor Yeah, I noticed. Looked through the logs on one of the failed jobs. This seems to be one of the culprits.

```
OAuthScopeInTokenResponseTest ++ 19:18:22,351 INFO  [org.keycloak.testsuite.oauth.OAuthScopeInTokenResponseTest] [OAuthScopeInTokenResponseTest] specifySingleScopeAsClientRoleTest() STARTED
OAuthScopeInTokenResponseTest ++ [33m19:18:22,446 WARN  [org.keycloak.services] (default task-105) KC-SERVICES0013: Failed authentication: java.lang.NullPointerException
OAuthScopeInTokenResponseTest ++ 	at org.keycloak.models.cache.infinispan.UserCacheSession.getNotBeforeOfUser(UserCacheSession.java:745)
OAuthScopeInTokenResponseTest ++ 	at org.keycloak.services.managers.AuthenticationManager.verifyIdentityToken(AuthenticationManager.java:1097)
OAuthScopeInTokenResponseTest ++ 	at org.keycloak.services.managers.AuthenticationManager.authenticateIdentityCookie(AuthenticationManager.java:662)
OAuthScopeInTokenResponseTest ++ 	at org.keycloak.authentication.authenticators.browser.CookieAuthenticator.authenticate(CookieAuthenticator.java:42)
```

Not sure why it's happening though.

**Comment by pedroigor** [comment]
I can give it a try. Will check before this week ends (soon).

**Comment by hardselius** [comment]
@pedroigor I think I've located the error. Will run the tests locally and push an update during the day,.

**Comment by pedroigor** [comment]
@hardselius, it looks good now. The only thing we could improve here is changes to tests as you could just extend those classes instead of including pairwise related methods there. But this is something I can change later and what you did is enough.

Let me wait a little bit for Marek's feedback on your changes to ```AuthenticationManager```. After that I'll merge your PR.

Thanks in advance for your contribution.

**Comment by mposolda** [comment]
@hardselius @pedroigor changes in AuthenticationManager looks good to me. Yeah, I guess that AuthenticationManager.verifyIdentityToken previously just didn't work for pairwise subjects..

**Review by pedroigor — APPROVED** [review]

**Pull request merged**